File: csharp/oelMVCS/Controller/ControllerCenter.cs
/********************************************************************
     Copyright (c) XTechCloud
     All rights reserved.
*********************************************************************/

using System.Collections.Generic;

namespace XTC.oelMVCS
{
    internal class ControllerCenter
    {
        public ControllerCenter(Board _board)
        {
            board_ = _board;
        }

        public Error Register(string _uuid, Controller.Inner _inner)
        {
            board_.getLogger()?.Info("register controller: {0}", _uuid);
            if (unitS_.ContainsKey(_uuid))
                return Error.NewAccessErr("controller {0} exists", _uuid);
            unitS_[_uuid] = _inner;
            return Error.OK;
        }

        public Error Cancel(string _uuid)
        {
            board_.getLogger()?.Info("cancel controller: {0}", _uuid);
            if (!unitS_.ContainsKey(_uuid))
                return Error.NewAccessErr("controller {0} not found", _uuid);
            unitS_.Remove(_uuid);
            return Error.OK;
        }

        public Controller.Inner? FindUnit(string _uuid)
        {
            Controller.Inner? inner;
            if (!unitS_.TryGetValue(_uuid, out inner))
                return null;
            return inner;
        }

        public void PreSetup()
        {
            board_.getLogger()?.Info("perSetup controllers");
            foreach (Controller.Inner inner in unitS_.Values)
            {
                board_.getLogger()?.Debug("perSetup controller: {0}", inner.getUnit().getUID());
                inner.PreSetup();
            }
        }

        public void Setup()
        {
            board_.getLogger()?.Info("setup controllers");
            foreach (Controller.Inner inner in unitS_.Values)
            {
                board_.getLogger()?.Debug("setup controller: {0}", inner.getUnit().getUID());
                inner.Setup();
            }
        }

        public void PostSetup()
        {
            board_.getLogger()?.Info("postSetup controllers");
            foreach (Controller.Inner inner in unitS_.Values)
            {
                board_.getLogger()?.Debug("postSetup controller: {0}", inner.getUnit().getUID());
                inner.PostSetup();
            }
        }

        public void PreDismantle()
        {
            board_.getLogger()?.Info("perDismantle controllers");
            foreach (Controller.Inner inner in unitS_.Values)
            {
                board_.getLogger()?.Debug("perDismantle controller: {0}", inner.getUnit().getUID());
                inner.PreDismantle();
            }
        }

        public void Dismantle()
        {
            board_.getLogger()?.Info("dismantle controllers");
            foreach (Controller.Inner inner in unitS_.Values)
            {
                board_.getLogger()?.Debug("dismantle controller: {0}", inner.getUnit().getUID());
                inner.Dismantle();
            }
        }

        public void PostDismantle()
        {
            board_.getLogger()?.Info("postDismantle controllers");
            foreach (Controller.Inner inner in unitS_.Values)
            {
                board_.getLogger()?.Debug("postDismantle controller: {0}", inner.getUnit().getUID());
                inner.PostDismantle();
            }
        }

        private Dictionary<string, Controller.Inner> unitS_ = new Dictionary<string, Controller.Inner>();

        private Board board_;
    }
}//namespace

File: python/oelMVCS/Controller/ControllerCenter.py
# -*- coding: utf-8 -*-

from typing import Dict, NewType

class ControllerCenter:
    """控制层中心
    """

    def __init__(self, _board:Board):
        self.__controllers : Dict[str,Controller.Inner] = {}
        self.__board :Board = _board

    def Register(self, _uuid:str, _inner:Controller.Inner) -> Error:
        self.__board.getLogger().Info("register controller {}", _uuid)
        if (_uuid in __controllers.keys()):
            return Error.NewAccessErr("controller {} exists", _uuid)
        self.__controllers[_uuid] = _inner
        return Error.OK

    def Cancel(self, string _uuid) -> Error:
        self.__board.getLogger().Info("cancel controller {}", _uuid)
        if (not _uuid in self.__controllers.keys()):
            return Error.NewAccessErr("controller {} not found", _uuid)
        self.__controllers.remove(_uuid)
        return Error.OK

    def FindController(self, _uuid:str) -> Controller.Inner:
        return self.__controllers[_uuid]

    def Setup(self):
        for inner in self.__controllers.values():
            inner.Setup(self.__board)

    def Dismantle(self):
        for inner in self.__controllers.values():
            inner.Dismantle();
